core/core-impl/src/main/java/ai/timefold/solver/core/impl/heuristic/selector/move/generic/list/kopt/TwoOptListMove.java

```diff
@@ -75,6 +77,7 @@ public TwoOptListMove(ListVariableDescriptor<Solution_> variableDescriptor,
                 shift = 0;
             }
         } else {
+            entityFirstUnpinnedIndex = 0;
```

Comment:
This looks fishy.
How have you guaranteed that the first unpinned index is 0 here?

Reply:
This path is for tail swaps (two different entities) and thus does not do flipping (and won't wrap around); will add a comment.

Reply:
Still... the first unpinned index could be `> 0`, no?
The tail must start after the pin.

Reply:
Yes, and if it does not, that a bug in your code, since it just picks two random values from valueSelectors, and valueSelectors should not return pinned values.

Reply:
I'm confused. My code guarantees two things:

1. Fully pinned entities will not be returned by entity selectors.
2. Pinned values will not be returned by value selectors.

But in this case, neither of the two is in play. It says here that the pin index is `0`, regardless of whether the pin actually is `0`. If you move the thing at index `0` and the pin is `1`, you have moved a pinned thing - value selectors can not save you from that.

Which is why I'm asking how is it possible that you can _know_ that 0 is safe here. Because I probably don't see some connection that you are seeing.

Reply:
Ah, I think I understand... you're saying that the value is *relative* to the pin, not the absolute value of the pin. So `0` means the first value after the pin.

It would've been clearer if the variable said that, or perhaps a comment.

Reply:
Because TwoOptListMove is two moves implemented by the same class:

1. If the entities both values are on are the same, it a SublistFlipMove, which needs firstUnpinnedIndex so it can flip correctly if the second value comes before the first value
2. If the values are on different entities, it a TailSwapMove, which does not need or use firstUnpinnedIndex